kernel: Rm ShutdownRequested and AbortNode from validation code.

sedited opened the pull request:
Get rid of all `ShutdownRequested` calls in validation code by introducing an interrupt object that applications can use to cancel long-running kernel operations.

Replace all `AbortNode` calls in validation code with new fatal error and flush error notifications so kernel applications can be notified about failures and choose how to handle them. 

---

This pull request is part of the `libbitcoinkernel` project https://github.com/bitcoin/bitcoin/issues/27587 https://github.com/orgs/bitcoin/projects/3 and more specifically its "Step 2: Decouple most non-consensus code from libbitcoinkernel".

The pull request mostly allows dropping the kernel dependency on shutdown.cpp. The only dependency left after this is a `StartShutdown` call which will be removed in followup PR https://github.com/bitcoin/bitcoin/pull/27711. This PR also drops the last reference to the `uiInterface` global in kernel code. The process of moving the `uiInterface` out of the kernel was started in https://github.com/bitcoin/bitcoin/pull/27636.

This pull request contains a subset of patches originally proposed in #27711. It will be part of a series of changes required to make handling of interrupts (or in other words the current shutdown procedure) in the kernel library more transparent and less reliable on global mutable state. The set of patches contained here was originally proposed by @ryanofsky [here](https://github.com/bitcoin/bitcoin/pull/27711#issuecomment-1580779869).

sedited committed:
kernel: Add fatalError method to notifications

FatalError replaces what previously was the AbortNode function in
shutdown.cpp.

This commit is part of the libbitcoinkernel project and further removes
the shutdown's and, more generally, the kernel library's dependency on
interface_ui with a kernel notification method. By removing interface_ui
from the kernel library, its dependency on boost is reduced to just
boost::multi_index. At the same time it also takes a step towards
de-globalising the interrupt infrastructure.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>
Co-authored-by: TheCharlatan <seb.kung@gmail.com>

sedited committed:
scripted-diff: Rename FatalError to FatalErrorf

This is done in preparation for the next commit where a new FatalError
function is introduced. FatalErrorf follows common convention to append
'f' for functions accepting format arguments.

-BEGIN VERIFY SCRIPT-
sed -i 's/FatalError/FatalErrorf/g' $( git grep -l 'FatalError')
-END VERIFY SCRIPT-

sedited committed:
kernel: Pass interrupt reference to chainman

This and the following commit seek to decouple the libbitcoinkernel
library from the shutdown code. As a library, it should it should have
its own flexible interrupt infrastructure without relying on node-wide
globals.

The commit takes the first step towards this goal by de-globalising
`ShutdownRequested` calls in kernel code.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>
Co-authored-by: TheCharlatan <seb.kung@gmail.com>

sedited committed:
util: Add SignalInterrupt class and use in shutdown.cpp

This change helps generalize shutdown code so an interrupt can be
provided to libbitcoinkernel callers. This may also be useful to
eventually de-globalize all of the shutdown code.

Co-authored-by: Russell Yanofsky <russ@yanofsky.org>
Co-authored-by: TheCharlatan <seb.kung@gmail.com>

ryanofsky reviewed (approved):
Code review ACK d60ab3b5099d3a348f3122b5d3207ac12f4a751c. Overall looks great! Feel free to ignore most of comments I left. The only thing I think really should be addressed is the two `BlockManager::Flush` calls aborting without returning errors. I think this is confusing and potentially could lead to bugs, but since the changes here don't actually affect the way the code run in practice, it doesn't need to block the PR.

ryanofsky commented:
> > The only thing I think really should be addressed is the two BlockManager::Flush calls aborting without returning errors.
> 
> I was about to open a separate PR from this branch here: https://github.com/TheCharlatan/bitcoin/tree/kernelReturnOnAbort

Great! I also opened #27862 early which overlaps with the branch a little. I think between the branch and PR all the cases where errors are not currently returned are covered.

TheCharlatan commented:
Rebased d60ab3b5099d3a348f3122b5d3207ac12f4a751c -> 3e60c02ef3759e16a762bd193e0b9291bf46b73c ([kernelInterrupt_0](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_0) -> [kernelInterrupt_1](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_1), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_0..kernelInterrupt_1)).
* Added `src/node/abort.*` to allow both the notification code and `index/base.cpp` to use a common abort definition as introduced in https://github.com/bitcoin/bitcoin/commit/3c06926cf21dcca3074ef51506f556b2286c299b
* De-globalized the `exit_status` in `shutdown.*`. My current approach touches quite a few lines though and seems a bit unclean, suggestions for alternative approaches are appreciated.

Updated 3e60c02ef3759e16a762bd193e0b9291bf46b73c -> 39d9b5144410c15385ee48bc886e0dea36f34b6a ([kernelInterrupt_1](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_1) -> [kernelInterrupt_2](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_2), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_1..kernelInterrupt_2)).
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226706890), cleaning up includes.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226742797), renaming sleep to wait.
* Addressed @ryanofsky's comments ([1](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226757927), [2](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226772621), [3](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226776471)), avoiding behavior changes around atomic memory acquisition and release.
* Addressed @ryanofsky's [comment](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226800991), removing a field that was accidentally carried over from the parent PR.

ryanofsky reviewed (approved):
Code review ACK 39d9b5144410c15385ee48bc886e0dea36f34b6a. I left a suggestion to reduce size of changes to index code, and I still think it is inappropriate to treat flush errors as fatal errors, and would prefer to add a separate flush error method (https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226847296), but I won't get hung up on this.

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/27861#issuecomment-1589076014

> * De-globalized the `exit_status` in `shutdown.*`. My current approach touches quite a few lines though and seems a bit unclean, suggestions for alternative approaches are appreciated.

I'm actually not sure what this is referring to. `exit_status` is not referenced in `shutdown.cpp` at all after this PR: https://github.com/TheCharlatan/bitcoin/blob/39d9b5144410c15385ee48bc886e0dea36f34b6a/src/shutdown.cpp.

Also, after this PR `shutdown.cpp` is just a thin wrapper around the `kernel::Context::interrupt` object, so we could pretty easily delete `shutdown.cpp` entirely and update callers to use the object directly.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/27861#pullrequestreview-1477669211

> and I still think it is inappropriate to treat flush errors as fatal errors, and would prefer to add a separate flush error method (https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226847296), but I won't get hung up on this.

ACK, I'll revert to using `AbortNode` then for now, that should keep any potential churn minimal. Once https://github.com/bitcoin/bitcoin/pull/27866 has received sufficient review, we can always decide to add special cases (e.g. flush error notifications) afterwards.

hebasto commented:
Concept ACK.

TheCharlatan commented:
Re https://github.com/bitcoin/bitcoin/pull/27861#pullrequestreview-1477669211

> and I still think it is inappropriate to treat flush errors as fatal errors, and would prefer to add a separate flush error method (https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226847296)

Just to ensure I am understanding you here, would the `node::KernelNotifications` for this extra notification match the implementation of the `fatalError` method for now (and later be adapted as needed), or do you want it to just log?

ryanofsky commented:
> would the `node::KernelNotifications` for this extra notification match the implementation of the `fatalError` method for now (and later be adapted as needed), or do you want it to just log?

It should definitely call AbortNode and not change current behavior. The point of the suggestion is send fatal error notiifications in a consistent way and not introduce misleading code that says an error will be returned separately when it won't be.

ryanofsky reviewed (approved):
Code review ACK 315915b43082a4c43bec3ba8e826abc7280b6cbc. Only change since last review is simplifying changes to index code as suggested earlier.

I still think it would be nicer if `FlushUndoFile` and `FlushBlockFile` sent flush error notifications instead of fatal error notifications to avoid lumping together errors that mean different things and are not handled the same way (https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1226847296). But this would just be for clarity. Behavior would be the same either way so it's not that important.

sedited committed:
kernel: Add flushError method to notifications

This is done in addition with the following commit. Both have the goal
of getting rid of direct calls to AbortNode from kernel code. This extra
flushError method is added to notify specifically about errors that
arrise when flushing (syncing) block data to disk. Unlike other
instances, the current calls to AbortNode in the blockstorage flush
functions do not report an error to their callers.

This commit is part of the libbitcoinkernel project and further removes
the shutdown's and, more generally, the kernel library's dependency on
interface_ui with a kernel notification method. By removing interface_ui
from the kernel library, its dependency on boost is reduced to just
boost::multi_index. At the same time it also takes a step towards
de-globalising the interrupt infrastructure.

TheCharlatan commented:
Thank you for bearing with me @ryanofsky and the review @hebasto,

Updated 315915b43082a4c43bec3ba8e826abc7280b6cbc -> 8654e835dd12a96b98eb554da48e342270e7279b ([kernelInterrupt_3](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_3) -> [kernelInterrupt_4](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_4), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_3..kernelInterrupt_4)).
* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1231072969), removing source file included by accident in makefile sources list.
* Addressed @hebasto's [comment](https://github.com/bitcoin/bitcoin/pull/27861#discussion_r1231090091), improving documentation formatting.
* Adopted @ryanofsky's [suggestions](https://github.com/bitcoin/bitcoin/pull/27861#pullrequestreview-1482367554), added a commit for separate notification handling of fatal flush errors.

Rebased 8654e835dd12a96b98eb554da48e342270e7279b -> 5155ce6114e199545a6ebdc77425dedc20233610 ([kernelInterrupt_4](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_4) -> [kernelInterrupt_5](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_5), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_4..kernelInterrupt_5)).
* Hopefully this fixes the CI failure.

evansmj commented on src/util/signalinterrupt.h:
```diff
@@ -0,0 +1,52 @@
+
+namespace util {
+/**
+ * Helper class that allows manages an interrupt flag, and allows a thread or
```
`Helper class that [allows] [manages] an interrupt flag`
Was this meant to be `Helper class that manages an interrupt flag`?

evansmj commented on src/util/signalinterrupt.h:
```diff
@@ -0,0 +1,52 @@
+ * signal to interrupt another thread.
+ *
+ * This class is safe to be used in a signal handler. If sending an interrupt
+ * from a signal handler is not neccessary, the more lightweight \ref
```
`neccessary` can be updated to `necessary`

evansmj commented on src/util/threadinterrupt.h:
```diff
@@ -12,11 +12,18 @@
+ * until reset.
+ *
+ * This class should not be used in a signal handler. If sending an interrupt
+ * from a signal handler is neccessary, the \ref SignalInterrupt class can be
```
`necessary`

evansmj reviewed (commented):
just a few typos i found

TheCharlatan commented:
Updated 5155ce6114e199545a6ebdc77425dedc20233610 -> e0653c2223f8d6d74645c9b5742de1c069d8c3a7 ([kernelInterrupt_5](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_5) -> [kernelInterrupt_6](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_6), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_5..kernelInterrupt_6)).
* Fixed typos as suggested by @evansmj

ryanofsky commented on src/kernel/notifications_interface.h:
```diff
@@ -32,6 +33,12 @@ class Notifications
+    //! The fatal error notification is sent to notify the user and start
+    //! shutting down if an error happens in kernel code that can't be recovered
+    //! from. After this notification is sent, whatever function triggered the
+    //! error should also return an error code or raise an exception.
```
In commit "kernel: Add fatalError method to notifications" (e0653c2223f8d6d74645c9b5742de1c069d8c3a7)

Maybe add "This notification can be used to log the error, or notify the user, trigger an application shutdown to prevent causing more errors."

ryanofsky reviewed (approved):
Code review ACK e0653c2223f8d6d74645c9b5742de1c069d8c3a7. Thanks for the updates! Just suggested a few small changes to comments

TheCharlatan commented:
Updated e0653c2223f8d6d74645c9b5742de1c069d8c3a7 -> 0c2fd16275ae078e33d71b0c9da8034075998553 ([kernelInterrupt_6](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_6) -> [kernelInterrupt_7](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_7), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_6..kernelInterrupt_7)).
* Adjusted and simplified documentation as suggested by @ryanofsky 

Rebased 0c2fd16275ae078e33d71b0c9da8034075998553 -> 001acf37dff40688d82c3b1a7ff9ed3901addb33 ([kernelInterrupt_7](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_7) -> [kernelInterrupt_8](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_8), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_7..kernelInterrupt_8)).
* Fixed conflicts with https://github.com/bitcoin/bitcoin/pull/27862

ryanofsky reviewed (approved):
Code review ACK 001acf37dff40688d82c3b1a7ff9ed3901addb33. This looks very good, and hopefully can be merged soon with more review.

To encourage review, I would maybe rewrite the PR title and description to be a little less abstract, and describe the benefits concretely. For title, would suggest "kernel: Remove all `ShutdownRequested` and `AbortNode` calls from validation code." For description would suggest: "Replace all `AbortNode` calls in validation code with new fatal error and flush error notifications so kernel applications can be notified about failures and choose how to handle them. Also, get rid of all `ShutdownRequested` calls in validation code by introducing an interrupt object that applications can use to cancel long-running kernel operations. This PR mostly allows dropping the kernel dependency on `shutdown.cpp`. The only dependency left after this is a `StartShutdown` call which will be removed in followup PR #27711. This PR also drops the last reference to the `uiInterface` global in kernel code."

Only changes since last review are rebase and some documentation updates. It does need another rebase, though, because there is a new minor conflict with #27896

TheCharlatan commented:
Rebased 001acf37dff40688d82c3b1a7ff9ed3901addb33 -> 6eb33bd0c21b3e075fbab596351cacafdc947472 ([kernelInterrupt_8](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_8) -> [kernelInterrupt_9](https://github.com/TheCharlatan/bitcoin/commits/kernelInterrupt_9), [compare](https://github.com/TheCharlatan/bitcoin/compare/kernelInterrupt_8..kernelInterrupt_9)).
* Fixed conflict with #27896

hebasto commented on src/kernel/context.cpp:
```diff
@@ -26,6 +29,8 @@ Context::Context()
 Context::~Context()
 {
     ECC_Stop();
+    assert(g_context);
```
Would it be more safer to move the assertion to the top of the desctuctor's code?

hebasto reviewed (commented):
ACK 6eb33bd0c21b3e075fbab596351cacafdc947472, I have reviewed the code and it looks OK.

ryanofsky reviewed (approved):
Code review ACK 6eb33bd0c21b3e075fbab596351cacafdc947472. No changes since last review other than rebase.

If I can make one more suggestion about PR the description, I would suggest to make "Replace all AbortNode calls..." and "Get rid of all ShutdownRequested calls.." the first and main paragraph, and put everything else that's there below it separated by a horizontal line. I think the other text which is abstract and relies on explanations in linked PRs is going to be intimidating and not to make much sense to people who aren't following the libbitcoinkernel project closely, even if they are familiar with this code.

I think the PR makes sense by itself and the "Replace all AbortNode calls..." and "Get rid of all ShutdownRequested calls.."  text describes the changes and immediate benefits. If reviewers want to go down the rabbit hole and read about libbitcoinkernel goals and phases, that's great, but it shouldn't be a prerequisite for understanding and reviewing the actual code.

---

EDIT: When first posted this comment I didn't realize you broke apart AbortNode and ShutdownRequested explanations into separate paragraphs and added the "As a standalone library..." preamble. Would also drop the preamble or move it to the end, because I think the point about code dependencies is incidental, and then main benefit this PR provides are interfaces for applications using the kernel library to:

1. Cancel long running operations
2. Receive notifications about errors

ryanofsky commented:
If anyone else wants to review: this PR has two acks and is a pure refactoring and I think could be merged if there is another reviewer.

achow101 commented:
light ACK 6eb33bd0c21b3e075fbab596351cacafdc947472

I'm not particularly experienced with this area of the code, so I can't really comment on approach taken here. However mechanically, the code looks find and correct to me and seems to achieve the stated objectives.

maflcko reviewed (commented):
Concept ACK, left some questions.



Concept ACK 6eb33bd0c21b3e075fbab596351cacafdc947472 🗼

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: Concept ACK 6eb33bd0c21b3e075fbab596351cacafdc947472 🗼
NeAejYNkmlMXartwqcjqc2A5dHtRJ8FfNWSSKXnDRNWaaFh+q2iMhmXxlc2pMsdyhs4X2otttjtOoOTRLqrhAQ==
```

</details>

ryanofsky commented on src/util/signalinterrupt.h:
```diff
@@ -0,0 +1,52 @@
+    explicit operator bool() const;
+    void operator()();
+    void reset();
+    void wait();
```
> Also, it would be good to explain why in the following commits it is fine to then ignore the exceptions and no longer catch them.

I probably just missed something, but there shouldn't be places in the following commits where exceptions are no longer caught. This should just be a refactoring.

I agree though that throwing undocumented exceptions is not good. And probably getting rid of exceptions and using nodiscard return values would be better than documenting them. The reason these are using exceptions is I was trying to make the `SignalInterrupt` interface consistent with the `CThreadInterrupt` interface, and there wasn't a good way to return an error from the constructor.

I will follow up with this in #28051
